File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/PlanGrouping.kt
package ai.jolli.jollimemory.core

/**
 * PlanGrouping — Kotlin port of vscode/src/util/PlanGrouping.ts (the push-relevant subset).
 *
 * A plan committed to a commit is archived under a slug that embeds the commit
 * hash (`<base-slug>-<shortHash>`). Squash consolidation hoists every source
 * commit's plans into the consolidated commit, so the same logical plan appears
 * once per source commit — same title, different slug. These helpers collapse
 * those snapshots by base name and pick the latest, so the push path never mints
 * duplicate same-named Space documents.
 */
object PlanGrouping {

    private val HASH_SUFFIX = Regex("-[0-9a-f]{8}$")

    /**
     * Strips a trailing archived commit-hash suffix (`-<8 hex>`) to get the base name.
     * Committed snapshots (`refactor-auth-a1b2c3d4`) and an uncommitted base
     * (`refactor-auth`) collapse to the same key.
     */
    fun planBaseKey(slug: String): String = slug.replace(HASH_SUFFIX, "")

    /**
     * Compares two plans newest-first by `updatedAt`, tiebroken by `slug` so the
     * order is deterministic.
     */
    private fun byUpdatedAtDesc(a: PlanReference, b: PlanReference): Int {
        if (a.updatedAt != b.updatedAt) return if (a.updatedAt < b.updatedAt) 1 else -1
        return a.slug.compareTo(b.slug)
    }

    /**
     * Returns exactly one plan per base name — the latest snapshot — preserving the
     * newest-first order. Used to avoid pushing duplicate same-named documents to Jolli.
     *
     * Same-named plans share an identical server push identity (the slug is NOT sent),
     * so `jolliPlanDocId` is the only thing that tells the server to UPDATE rather than
     * CREATE. When a previously pushed older snapshot carries the docId but the latest
     * snapshot does not, the latest inherits that docId/url so the push updates the
     * existing article instead of creating a duplicate.
     */
    fun latestPlanPerName(plans: List<PlanReference>): List<PlanReference> {
        val sorted = plans.sortedWith(::byUpdatedAtDesc)
        // Newest already-pushed docId/url per base name (first hit wins = newest).
        val pushedDoc = HashMap<String, Pair<Int, String?>>()
        for (plan in sorted) {
            val key = planBaseKey(plan.slug)
            if (plan.jolliPlanDocId != null && !pushedDoc.containsKey(key)) {
                pushedDoc[key] = plan.jolliPlanDocId to plan.jolliPlanDocUrl
            }
        }
        val seen = HashSet<String>()
        val result = ArrayList<PlanReference>()
        for (plan in sorted) {
            val key = planBaseKey(plan.slug)
            if (!seen.add(key)) continue
            if (plan.jolliPlanDocId == null) {
                val inherited = pushedDoc[key]
                if (inherited != null) {
                    result.add(plan.copy(jolliPlanDocId = inherited.first, jolliPlanDocUrl = inherited.second))
                    continue
                }
            }
            result.add(plan)
        }
        return result
    }
}

File: cli/src/core/push/PlanGrouping.ts
/**
 * PlanGrouping — same-name plan-snapshot collapsing, shared by the plan context
 * kind and by `JolliMemoryPushOrchestrator` (which re-exports these so its
 * existing importers, including the ide-bridge, are unaffected).
 *
 * Moved out of the orchestrator when the push path became table-driven: the plan
 * definition (in `kinds/index.ts`) needs them, and the orchestrator imports the
 * registry, so leaving them in the orchestrator would close an import cycle.
 * Behaviour is unchanged — these are the original implementations verbatim.
 */

import type { PlanReference } from "../../Types.js";
import { REF_HASH_SUFFIX } from "../RefMerge.js";

/**
 * Strips a trailing archived commit-hash suffix (`-<8 hex>`) to get the base
 * name. Committed snapshots (`refactor-auth-a1b2c3d4`) and an uncommitted base
 * (`refactor-auth`) collapse to the same key.
 *
 * Shares `REF_HASH_SUFFIX` with RefMerge's `baseKeyOf.plan` rather than re-spelling
 * the pattern: this is the same base key the amend hoist dedupes by, and a drift
 * between the two would have the push path group snapshots the summary kept apart.
 */
export function planBaseKey(slug: string): string {
	return slug.replace(REF_HASH_SUFFIX, "");
}

/**
 * Compares two plans newest-first by `updatedAt`, tiebroken by `slug` so the
 * order is deterministic across repeated calls.
 */
export function byUpdatedAtDesc(a: PlanReference, b: PlanReference): number {
	if (a.updatedAt !== b.updatedAt) {
		return a.updatedAt < b.updatedAt ? 1 : -1;
	}
	return a.slug < b.slug ? -1 : a.slug > b.slug ? 1 : 0;
}

/**
 * Returns exactly one plan per base name — the latest snapshot — preserving the
 * newest-first order. Used to avoid pushing duplicate same-named documents to
 * Jolli.
 *
 * Same-named plans share an identical server push identity (same title, branch,
 * relativePath, commit — the slug is NOT sent), so `jolliPlanDocId` is the only
 * thing that tells the server to UPDATE rather than CREATE. When a previously
 * pushed older snapshot carries the docId but the latest snapshot does not, the
 * latest inherits that docId/url so the push updates the existing article
 * instead of creating a duplicate (which the server rejects → push failure).
 */
export function latestPlanPerName(plans: ReadonlyArray<PlanReference>): ReadonlyArray<PlanReference> {
	const sorted = [...plans].sort(byUpdatedAtDesc);
	// Newest already-pushed docId/url per base name (first hit wins = newest). The
	// URL rides with the docId so the reuse gate downstream (`canReuseDocId`, which
	// reads the URL's origin) can tell which backend the inherited id belongs to.
	const pushedDoc = new Map<string, { docId: number; url: string | undefined }>();
	for (const plan of sorted) {
		const key = planBaseKey(plan.slug);
		if (plan.jolliPlanDocId !== undefined && !pushedDoc.has(key)) {
			pushedDoc.set(key, { docId: plan.jolliPlanDocId, url: plan.jolliPlanDocUrl });
		}
	}
	const seen = new Set<string>();
	const result: PlanReference[] = [];
	for (const plan of sorted) {
		const key = planBaseKey(plan.slug);
		if (seen.has(key)) {
			continue;
		}
		seen.add(key);
		if (plan.jolliPlanDocId === undefined) {
			const inherited = pushedDoc.get(key);
			if (inherited) {
				result.push({
					...plan,
					jolliPlanDocId: inherited.docId,
					jolliPlanDocUrl: inherited.url,
				});
				continue;
			}
		}
		result.push(plan);
	}
	return result;
}
